ingest/cell_metadata.py

```diff
@@ -1,39 +1,94 @@
-#! /usr/bin/python
-
-from collections import defaultdict
-from itertools import islice
-
-
-class CellMetadata:
-    """A class used to represent cell metadata.
-
-    Attributes
-    ----------
-    headers : list of strings
-        list of metadata names in metadata file
-    metadata_types : list of strings
-        list of metadata type designations from metadata file
-    annotation_type : list of strings
-        list of valid metadata type designations
-    errors : dict
-        dict of collected validation errors and warnings
-
-    """
-    def __init__(self, file_path):
-        """
-        Parameters
-        ----------
-        filepath : str
-            The name of the metadata file
-
-        """
-        self.file = open(file_path, 'r')
-        self.headers = self.file.readline().rstrip('\n').split('\t')
-        self.metadata_types = self.file.readline().rstrip('\n').split('\t')
-        self.annotation_type = ['group', 'numeric']
-        self.errors = defaultdict(list)
-        self.ontology = defaultdict(lambda: defaultdict(set))
-        self.type = defaultdict(list)
+"""Module for ingesting cell metadata files
+
+DESCRIPTION
+Module provides extract and transform functions for cell metadata files.
+Text, CSV, and TSV files are supported.
+
+PREREQUISITES
+Must have python 3.6 or higher.
+"""
+import copy
+import json
+import os
+from typing import Dict, Generator, List, Tuple, Union
+
+from ingest_files import IngestFiles
+
+
+class CellMetadata(IngestFiles):
+    ALLOWED_FILE_TYPES = ['text/csv',
+                          'text/plain', 'text/tab-separated-values']
+
+    def __init__(self, file_path, file_id: str = None, study_accession: str = None):
+
+        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.headers = self.get_next_line(
+            increase_line_count=False)
+        self.metadata_types = self.get_next_line(
+            increase_line_count=False)
+        # unique values for group-based annotations
+        self.uniqueValues = []
+        self.cell_names = []
+        self.top_level_doc = self.create_documents(
+            file_path, file_id, study_accession)
+        self.data_subcollection = self.create_subdocuments()
+        self.errors = {}
+
+    def transform(self, row: List[str]) -> None:
+        """ Add data from cell metadata files into data model"""
+        for idx, column in enumerate(row):
+            if idx != 0:
+                # if annotation is numeric convert from string to float
+                if self.metadata_types[idx].lower() == 'numeric':
```

Comment:
Thank you for calling lower() on the header value!  I'm glad we're learning from previous regressions.